### PR TITLE
Adding parquet transcoding example

### LINKED ISSUE #15344: [FEA] Add Parquet transcoding to `cpp/examples`
**Is your feature request related to a problem? Please describe.**
Recently we added a libcudf example for [processing nested data](https://github.com/rapidsai/cudf/tree/branch-24.04/cpp/examples/nested_types) types. The [deduplication example](https://github.com/rapidsai/cudf/blob/branch-24.04/cpp/examples/nested_types/deduplication.cpp) uses a command line interface to receive a filename, perform some relational algebra and output basic timing data to the console. 

Let's add an example that performs parquet file transcoding. The example can read a Parquet file that contains a single column, and then write it using a specified encoding and compression, and then read the file again. Finally, the example can confirm the data is the same between the first and second reads.

**Describe the solution you'd like**

Here is an snippet showing how the example might be called:
`./parquet_io ~/in.pq ~/out.pq DELTA_BYTE_ARRAY ZSTD`
where the parameters represent input_filepath, output_filepath, column_encoding and compression_type. [Valid encodings](https://github.com/rapidsai/cudf/blob/f9ac4277f50163a7da2006460034aa3e45c8744e/cpp/include/cudf/io/types.hpp#L105) include  DICTIONARY, PLAIN, DELTA_BINARY_PACKED, DELTA_LENGTH_BYTE_ARRAY, DELTA_BYTE_ARRAY, plus ([soon to be](https://github.com/rapidsai/cudf/pull/15311)) BYTE_STREAM_SPLIT. 

The example will print to console the time elapsed for (1) the initial read, (2) the write, and (3) the second read.

The example does not need to verify is the requested encoding is valid (e.g. string column with DELTA_BINARY_PACKED or int64 column with DELTA_BYTE_ARRAY). Let's only operate on the first column to keep things simple. 

**Describe alternatives you've considered**
Use a cuDF-python example instead, but I'd rather have a C++ example for each feature that we write a blog about.

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description
This PR adds a new example `parquet_io` to `libcudf/cpp/examples` instrumenting reading and writing parquet files with different column encodings (same for all columns for now) and compressions to close #15344. The example maybe elaborated and/or evolved as needed. #15348 should be merged before this PR to get all CMake updates needed to successfully build and run this example. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by GregoryKimball** [comment]
Thank you @mhaseeb123 for writing this example! It has made it easy for me to do a ton of testing with Kaggle data 🤩 

There are a few benchmarking items that we might want to address in the example, maybe with a comment, maybe with code change.. up to you.

1. startup time on first read. I believe we do nvcomp and cufile loading on the first read, so it takes more like 3s instead of ~30-300 ms. We might want to pre-load the extra libraries or else leave the first read untimed.
2. RMM pool growth. I believe the example uses a pool and then this grows during the first read, adding to runtime. We might choose to set a default, or leave the first read untimed.
3. OS cache clearing. Currently the example will time the second read optimistically, at least if the file is in the OS cache. I'm not sure if written files are always in the OS cache or if they have to be read once first. We might want to investigate to see if cache clearing before  the second read has any impact on read time.
4. Allowing optional size metadata to be written. We added page size metadata to the writer #14000 and reader #14973. Perhaps we could enable this option via a CLI parameter.

**Comment by mhaseeb123** [comment]
Thank you @GregoryKimball for the feedback. I am thinking of not timing the first read to avoid both RMM pool growth, and nvcomp and cufile pitfalls. I will add a CLI option to clear OS cache before second and subsequent reads plus another CLI option for optional size metadata soon.

**Comment by GregoryKimball** [comment]
Thank you @mhaseeb123 for discussing the example. I think we will address 1,2,3 above by untiming the first read, and then resolve 4 by adding a writer option setting the sizes metadata to OFF

**Comment by mhaseeb123** [comment]
Hi reviewers @vuule @vyasr. Just pinging you as the PR is now ready to be reviewed

File: cpp/examples/parquet_io/parquet_io.hpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  auto duration_us() { return std::chrono::duration_cast<micros>(end_time - start_time).count(); }
+  auto duration_ms() { return std::chrono::duration_cast<millis>(end_time - start_time).count(); }
+
+  void print_elapsed_micros() { std::cout << "Elapsed Time: " << duration_us() << "us\n\n"; }
+  void print_elapsed_millis() { std::cout << "Elapsed Time: " << duration_ms() << "ms\n\n"; }
+
+ private:
+  using time_point_t = std::chrono::time_point<std::chrono::high_resolution_clock>;
+  time_point_t start_time;
+  time_point_t end_time;
```
Comment (vuule) [inline cpp/examples/parquet_io/parquet_io.hpp:161]: fairly nitpicky given that this is sample code, but I feel like `Timer` could be safer/simpler to use.
```suggestion
  Timer() { reset(); }

  void reset() { start_time = std::chrono::high_resolution_clock::now(); }

  auto elapsed() { return (std::chrono::high_resolution_clock::now() - start_time).count(); }

  void print_elapsed_micros() { std::cout << "Elapsed Time: " << std::chrono::duration_cast<micros>(elapsed()) << "us\n\n"; }
  void print_elapsed_millis() { std::cout << "Elapsed Time: " << std::chrono::duration_cast<millis>(elapsed()) << "ms\n\n"; }

 private:
  using time_point_t = std::chrono::time_point<std::chrono::high_resolution_clock>;
  time_point_t start_time;
```

Comment (mhaseeb123) [reply cpp/examples/parquet_io/parquet_io.hpp]: All done. Thank you for the suggestion.

File: cpp/examples/parquet_io/parquet_io.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    std::cerr << e.what() << std::endl << std::endl;
+    std::cout << "Transcoding valid: false" << std::endl;
+    valid = false;
+  }
+
+  // No exception thrown, check for indices->size
+  if (valid) {
+    bool valid = indices->size() == 0;
+    std::cout << "Transcoding valid: " << std::boolalpha << valid << std::endl;
+  }
```
Comment (vuule) [inline cpp/examples/parquet_io/parquet_io.cpp:175]:
```suggestion
  try {
    // Left anti-join the original and transcoded tables
    // identical tables should not throw an exception and
    // return an empty indices vector
    auto const indices = cudf::left_anti_join(
      input->view(), transcoded_input->view(), cudf::null_equality::EQUAL, resource.get());
      
    // No exception thrown, check indices
    auto const valid = indices->size() == 0;
    std::cout << "Transcoding valid: " << std::boolalpha << valid << std::endl;
  } catch (std::exception& e) {
    std::cerr << e.what() << std::endl << std::endl;
    std::cout << "Transcoding valid: false" << std::endl;
  }
```

**Review by vuule — COMMENTED** [review]
few more suggestions, looks good otherwise :+1:

File: cpp/examples/parquet_io/parquet_io.hpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+#include <optional>
+#include <string>
+
+/**
+ * @brief Create memory resource for libcudf functions
+ *
+ * @param pool Whether to use a pool memory resource.
+ * @return Memory resource instance
+ */
+std::shared_ptr<rmm::mr::device_memory_resource> create_memory_resource(bool pool)
```
Comment (vuule) [inline cpp/examples/parquet_io/parquet_io.hpp:41]: nitpick: boolean parameter should be named is_xyz,something like
```suggestion
std::shared_ptr<rmm::mr::device_memory_resource> create_memory_resource(bool is_pool_used)
```

**Review by vuule — APPROVED** [review]
approved; just one optional suggestion

**Review by raydouglass — APPROVED** [review]

**Pull request merged**